File: 19.ts
// 19. Remove Nth Node From End of List
// https://leetcode.com/problems/remove-nth-node-from-end-of-list/

// Stack
// time complexity: O(n) || space complexity: O(n) (n: number of nodes)
function removeNthFromEnd1(head: ListNode | null, n: number): ListNode | null {
  if (!head) return head;

  const stack: ListNode[] = [];

  let pointer: ListNode | null = head;
  while (pointer) {
    stack.push(pointer);

    pointer = pointer.next;
  }

  const targetIndex = stack.length - n;

  if (targetIndex - 1 >= 0 && targetIndex + 1 <= stack.length - 1) {
    stack[targetIndex - 1].next = stack[targetIndex + 1];
  } else if (targetIndex - 1 < 0 && targetIndex + 1 <= stack.length - 1) {
    head = head.next;
  } else if (targetIndex - 1 >= 0 && targetIndex + 1 > stack.length - 1) {
    stack[targetIndex - 1].next = null;
  } else {
    head = null;
  }

  return head;
}

// Two Pointers
// time complexity: O(n) || space complexity: O(2) = O(1) (n: number of nodes)
function removeNthFromEnd2(head: ListNode | null, n: number): ListNode | null {
  if (!head) return head;

  let firstPointer = head;
  let secondPointer = head;

  for (let i = 0; i < n; i++) {
    firstPointer = firstPointer.next;
  }

  if (!firstPointer) return head.next;

  while (firstPointer.next) {
    secondPointer = secondPointer.next;
    firstPointer = firstPointer.next;
  }

  secondPointer.next = secondPointer.next.next;

  return head;
}

class ListNode {
  val: number;
  next: ListNode | null;
  constructor(val?: number, next?: ListNode | null) {
    this.val = val === undefined ? 0 : val;
    this.next = next === undefined ? null : next;
  }
}

File: LeetCode - Top Interview 150/19.py
# 19. Remove Nth Node From End of List
# https://leetcode.com/problems/remove-nth-node-from-end-of-list/


from typing import Optional


class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    # time complexity: O(m)
    def removeNthFromEnd1(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        stack = []

        start_node = ListNode(0, head)
        pointer = start_node.next

        while pointer:
            stack.append(pointer)

            pointer = pointer.next

        if len(stack) == 1:
            return None

        if n > 1 and n < len(stack):
            stack[-n - 1].next = stack[-n + 1]
        elif n == 1:
            stack[-2].next = None
        else:
            start_node.next = stack[1]

        return start_node.next

    # O O O O O & 3
    # O & 1
    # time complexity: O(m)
    def removeNthFromEnd2(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        start_node = ListNode(0, head)

        fast = head
        target = head
        target_before = start_node

        for _ in range(n):
            fast = fast.next

        while fast:
            fast = fast.next
            target = target.next
            target_before = target_before.next

        target_before.next = target.next

        return start_node.next
